cpp/memilio/mobility/graph.h

```diff
@@ -436,6 +442,112 @@ void print_graph(std::ostream& os, const Graph& g)
+public:
+    using NodeProperty = NodePropertyT;
+    using EdgeProperty = EdgePropertyT;
+
+    GraphBuilder() = default;
+    GraphBuilder(const size_t num_nodes, const size_t num_edges)
+    {
+        m_nodes.reserve(num_nodes);
+        m_edges.reserve(num_edges);
+    }
+
+    /**
+     * @brief Add a node to the GraphBuilder.
+     *
+     * The property of the node is constructed from arguments.
+     * @param id Id for the node.
+     * @tparam args Additional arguments for node construction.
+     */
+    template <class... Args>
+    void add_node(int id, Args&&... args)
+    {
+        m_nodes.emplace_back(id, std::forward<Args>(args)...);
+    }
+
+    /**
+     * @brief Add an edge to the GraphBuilder.
+     * 
+     * @param start_node_idx Id of start node
+     * @param end_node_idx Id of end node
+     * @tparam args Additional arguments for edge construction
+     */
+    template <class... Args>
+    void add_edge(size_t start_node_idx, size_t end_node_idx, Args&&... args)
+    {
+        assert(m_nodes.size() > start_node_idx && m_nodes.size() > end_node_idx);
+        m_edges.emplace_back(start_node_idx, end_node_idx, std::forward<Args>(args)...);
+    }
+
+    /**
+     * @brief Build the graph from the added nodes and edges.
+     * 
+     * Sorts the edges and optionally removes duplicate edges (same start and end node indices).
+     * @param make_unique If true, duplicate edges are removed. The first added edge is kept!
+     * @return Graph<NodePropertyT, EdgePropertyT> The constructed graph.
+     * @tparam NodeProperty The type of the node property.
+     * @tparam EdgeProperty The type of the edge property.
+     */
+    Graph<NodeProperty, EdgeProperty> build(bool make_unique = false)
+    {
+        sort_edges();
+        if (make_unique) {
+            remove_duplicate_edges();
+        }
+        Graph<NodeProperty, EdgeProperty> graph(std::move(m_nodes), std::move(m_edges));
+        return graph;
+    }
+
+private:
+    /**
+     * @brief Sort the edge vector of a graph.
+     */
+    void sort_edges()
+    {
+        std::stable_sort(m_edges.begin(), m_edges.end(), [](auto&& e1, auto&& e2) {
+            return e1.start_node_idx == e2.start_node_idx ? e1.end_node_idx < e2.end_node_idx
+                                                          : e1.start_node_idx < e2.start_node_idx;
+        });
+    }
+
+    /**
+     * @brief Remove duplicate edges from a sorted edge vector.
+     * 
+     * Copies all the unique edges to a new vector and replaces the original edge vector with it. Unique means that
+     * the start and end node indices are unique. Other edge properties are not checked and may get lost. Only the first 
+     * edge in the vector is kept.
+     */
+    void remove_duplicate_edges()
+    {
+        std::vector<Edge<EdgePropertyT>> unique_edges;
+        unique_edges.reserve(m_edges.size());
+        std::ranges::unique_copy(m_edges, std::back_inserter(unique_edges), [](auto&& e1, auto&& e2) {
```

Comment:
I'm still not entirely happy with this feature :D
In my opinion, we should at least log a warning (`log_warning`) when edges are deleted.